apps/rush-lib/src/logic/DependencyAnalyzer.ts

```diff
@@ -0,0 +1,192 @@
+  }
+
+  public static forRushConfiguration(rushConfiguration: RushConfiguration): DependencyAnalyzer {
+    if (!DependencyAnalyzer._dependencyAnalyzerByRushConfiguration) {
+      DependencyAnalyzer._dependencyAnalyzerByRushConfiguration = new Map();
+    }
+
+    let analyzer: DependencyAnalyzer | undefined =
+      DependencyAnalyzer._dependencyAnalyzerByRushConfiguration.get(rushConfiguration);
+    if (!analyzer) {
+      analyzer = new DependencyAnalyzer(rushConfiguration);
+      DependencyAnalyzer._dependencyAnalyzerByRushConfiguration.set(rushConfiguration, analyzer);
+    }
+
+    return analyzer;
+  }
+
+  public getAnalysis(variant?: string): IDependencyAnalysis {
+    // Use an empty string as the key when no variant provided. Anything else would possibly conflict
+    // with a variant created by the user
+    const variantKey: string = variant || '';
+    let analysis: IDependencyAnalysis | undefined = this._analysisByVariant.get(variantKey);
+    if (!analysis) {
+      const commonVersionsConfiguration: CommonVersionsConfiguration =
+        this._rushConfiguration.getCommonVersions(variant);
+
+      const allVersionsByPackageName: Map<string, Set<string>> = this._collectAllVersionsByPackageName(
+        commonVersionsConfiguration
+      );
+
+      const implicitlyPreferredVersionByPackageName: Map<string, string> =
+        this._getImplicitlyPreferredVersionByPackageName(
+          commonVersionsConfiguration,
+          allVersionsByPackageName
+        );
+
+      analysis = {
+        commonVersionsConfiguration,
+        implicitlyPreferredVersionByPackageName,
+        allVersionsByPackageName
+      };
+      this._analysisByVariant.set(variantKey, analysis);
+    }
+
+    return analysis;
+  }
+
+  /**
+   * Generates the {@link IDependencyAnalysis.allVersionsByPackageName} map.
+   *
+   * @remarks
+   * The result of this function is not cached.
+   */
+  private _collectAllVersionsByPackageName(
+    commonVersionsConfiguration: CommonVersionsConfiguration
+  ): Map<string, Set<string>> {
+    const allVersionsByPackageName: Map<string, Set<string>> = new Map();
+    const allowedAlternativeVersions: Map<
+      string,
+      ReadonlyArray<string>
+    > = commonVersionsConfiguration.allowedAlternativeVersions;
+    for (const project of this._rushConfiguration.projects) {
+      const dependencies: PackageJsonDependency[] = [
+        ...project.packageJsonEditor.dependencyList,
+        ...project.packageJsonEditor.devDependencyList
+      ];
+      for (const { name: dependencyName, version: dependencyVersion, dependencyType } of dependencies) {
+        if (dependencyType === DependencyType.Peer) {
+          // If this is a peer dependency, it isn't a real dependency in this context, so it shouldn't
+          // be included in the list of dependency versions.
+          continue;
+        }
+
+        const alternativesForThisDependency: ReadonlyArray<string> =
```

Comment:
It might be more efficient to do the checks against this collection in a post-processing step rather than while iterating projects. That way we only have to check each used version against the alternatives collection once.